Mathlib/Data/Sum/Basic.lean

```diff
@@ -201,6 +202,75 @@ section LiftRel
 #align sum.lift_rel.swap Sum.LiftRel.swap
 #align sum.lift_rel_swap_iff Sum.liftRel_swap_iff
 
+mk_iff_of_inductive_prop Sum.LiftRel Sum.liftRel_iff
+
+variable {r : α → γ → Prop} {s : β → δ → Prop} {x : Sum α β} {y : Sum γ δ}
+  {a : α} {b : β} {c : γ} {d : δ}
+
+theorem isLeft_eq_of_liftRel (h : LiftRel r s x y) : x.isLeft = y.isLeft := by
+  cases h <;> rfl
+
+theorem isRight_eq_of_liftRel (h : LiftRel r s x y) : x.isRight = y.isRight := by
+  cases h <;> rfl
+
+theorem isLeft_eq_of_liftRel_inl_fst (h : LiftRel r s (inl a) y) : y.isLeft :=
+  (isLeft_eq_of_liftRel h).symm
+
+theorem isLeft_eq_of_liftRel_inl_snd (h : LiftRel r s x (inl c)) : x.isLeft :=
+  isLeft_eq_of_liftRel h
+
+theorem isRight_of_liftRel_inr_fst (h : LiftRel r s (inr b) y) : y.isRight :=
+  (isRight_eq_of_liftRel h).symm
+
+theorem isRight_of_liftRel_inr_snd (h : LiftRel r s x (inr d)) : x.isRight :=
+  isRight_eq_of_liftRel h
+
+variable {τ κ : Type*} {f : τ → Sum α β} {g : κ → Sum γ δ} {t : τ} {k : κ}
+
+theorem isLeft_apply_of_inl_fst (h : LiftRel r s (inl a) (g k)) : (g k).isLeft :=
+  isLeft_eq_of_liftRel_inl_fst h
+
+theorem isLeft_apply_of_inl_snd (h : LiftRel r s (f t) (inl c)) : (f t).isLeft :=
+  isLeft_eq_of_liftRel_inl_snd h
+
+theorem isRight_apply_of_inr_fst (h : LiftRel r s (inr b) (g k)) : (g k).isRight :=
+  isRight_of_liftRel_inr_fst h
+
+theorem isRight_apply_of_inr_snd (h : LiftRel r s (f t) (inr d)) : (f t).isRight :=
+  isRight_of_liftRel_inr_snd h
+
+theorem isLeft_apply_eq_of_LiftRel_apply (h : LiftRel r s (f t) y) : (f t).isLeft = y.isLeft :=
+  isLeft_eq_of_liftRel h
+
+theorem isRight_apply_eq_of_LiftRel_apply (h : LiftRel r s x (g k)) :
+    (g k).isRight = x.isRight := (isRight_eq_of_liftRel h).symm
+
+theorem liftRel_of_left (h : ∃ a c, r a c ∧ x = inl a ∧ y = inl c) : LiftRel r s x y :=
+  (Sum.liftRel_iff ..).mpr (Or.inl h)
+
+theorem liftRel_of_right (h : ∃ b d, s b d ∧ x = inr b ∧ y = inr d) : LiftRel r s x y :=
+  (Sum.liftRel_iff ..).mpr (Or.inr h)
```

Comment:
Do you actually need these in practice? because you can just eliminate the existential and then this is just `LiftRel.inl` and `LiftRel.inr` again.

Reply:
No I think you're probably right, I just felt it might be useful to unwrap them. But I'm not wedded to them so I'm happy to remove if you think that's better.

Reply:
There might be an argument for having them but they're less clearly useful than the other lemmas in this PR so I think one of two things should happen now:
* either you go back to your project, try deleting the uses of these two lemmas and
  * either it's an awful mess and we keep them in this PR
  * or it's actually fine and we remove them from this PR
* or we remove the lemmas from this PR, merge it, you bump your project to the new mathlib and
  * either you can't remove the uses of the lemmas that didn't make it to mathlib and you open a new PR for them
  * or it's actually fine to only use what made it to mathlib and there's no need to do anything more

Reply:
Oh in the end I didn't use these in my project because I wanted to do something in less generality in my specific case. But I still want to introduce it into mathlib in the appropriate generality. I'll remove these for now.

Reply:
My strategy for integrating my project into mathlib is to do it in stages - I've factored out the stuff that is useful for reasons other than my specific definitions mostly, I'm just slowly PRing them all.